NoSuchElementException in SpecificationResourceCustomizer

## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Documentation issue or request
</code></pre>

## Description

This is from the [discussion](https://groups.google.com/g/syndesis/c/e2kwWxhCK4o/m/XnuznW3jAAAJ) on the mailing list. The relevant stack trace is this:

```
Caused by: java.util.NoSuchElementException: null
	at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:721) ~[na:1.8.0_191]
	at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:742) ~[na:1.8.0_191]
	at io.syndesis.connector.rest.swagger.SpecificationResourceCustomizer.updateSecuritySpecification(SpecificationResourceCustomizer.java:84) ~[connector-rest-swagger-1.10.0.jar!/:1.10.0]
	at io.syndesis.connector.rest.swagger.SpecificationResourceCustomizer.lambda$customize$0(SpecificationResourceCustomizer.java:47) ~[connector-rest-swagger-1.10.0.jar!/:1.10.0]
	at io.syndesis.integration.component.proxy.ComponentCustomizer.consumeOption(ComponentCustomizer.java:43) ~[integration-component-proxy-1.10.0.jar!/:1.10.0]
	at io.syndesis.connector.rest.swagger.SpecificationResourceCustomizer.customize(SpecificationResourceCustomizer.java:44) ~[connector-rest-swagger-1.10.0.jar!/:1.10.0]
	at io.syndesis.connector.rest.swagger.SpecificationResourceCustomizer.customize(SpecificationResourceCustomizer.java:37) ~[connector-rest-swagger-1.10.0.jar!/:1.10.0]
	at io.syndesis.integration.runtime.handlers.HandlerCustomizer.customizeComponent(HandlerCustomizer.java:67) ~[integration-runtime-1.10.0.jar!/:1.10.0]
	at io.syndesis.integration.runtime.handlers.ConnectorStepHandler.handle(ConnectorStepHandler.java:126) ~[integration-runtime-1.10.0.jar!/:1.10.0]
	at io.syndesis.integration.runtime.IntegrationRouteBuilder.configureFlow(IntegrationRouteBuilder.java:224) ~[integration-runtime-1.10.0.jar!/:1.10.0]
	at io.syndesis.integration.runtime.IntegrationRouteBuilder.configure(IntegrationRouteBuilder.java:143) ~[integration-runtime-1.10.0.jar!/:1.10.0]
```

And API document is this:

[tweeter-get-user.yaml.txt](https://github.com/syndesisio/syndesis/files/6082790/tweeter-get-user.yaml.txt)

--- a/app/common/util/src/main/java/io/syndesis/common/util/json/JsonUtils.java
+++ b/app/common/util/src/main/java/io/syndesis/common/util/json/JsonUtils.java
@@ -48,19 +48,21 @@ import io.syndesis.common.util.SyndesisServerException;
 public final class JsonUtils {
 
     private static final ObjectMapper OBJECT_MAPPER;
+    private static final ObjectMapper INCLUSIVE_MAPPER;
     private static final ObjectWriter OBJECT_WRITER;
     private static final ObjectReader OBJECT_READER;
 
     static {
-        OBJECT_MAPPER = new ObjectMapper()
-                .registerModules(new Jdk8Module(), new JavaTimeModule())
+        INCLUSIVE_MAPPER = new ObjectMapper()
+            .registerModules(new Jdk8Module(), new JavaTimeModule())
+            .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+            .enable(DeserializationFeature.READ_ENUMS_USING_TO_STRING)
+            .enable(SerializationFeature.WRITE_ENUMS_USING_TO_STRING)
+            .disable(JsonParser.Feature.AUTO_CLOSE_SOURCE)
+            .enable(MapperFeature.BLOCK_UNSAFE_POLYMORPHIC_BASE_TYPES);
+        OBJECT_MAPPER = INCLUSIVE_MAPPER.copy()
                 // keep using the deprecated method here in order to align with the jackson version (2.8.11) being used at integration runtime
-                .setPropertyInclusion(JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, JsonInclude.Include.NON_EMPTY))
-                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
-                .enable(DeserializationFeature.READ_ENUMS_USING_TO_STRING)
-                .enable(SerializationFeature.WRITE_ENUMS_USING_TO_STRING)
-                .disable(JsonParser.Feature.AUTO_CLOSE_SOURCE)
-                .enable(MapperFeature.BLOCK_UNSAFE_POLYMORPHIC_BASE_TYPES);
+                .setPropertyInclusion(JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, JsonInclude.Include.NON_EMPTY));
         OBJECT_READER = OBJECT_MAPPER.reader();
         OBJECT_WRITER = OBJECT_MAPPER.writer();
     }
@@ -122,7 +124,7 @@ public final class JsonUtils {
     }
 
     public static <T> T convertValue(final Object fromValue, final Class<T> toValueType) {
-        return OBJECT_MAPPER.convertValue(fromValue, toValueType);
+        return INCLUSIVE_MAPPER.convertValue(fromValue, toValueType);
     }
 
     /**

--- a/app/connector/rest-swagger/pom.xml
+++ b/app/connector/rest-swagger/pom.xml
@@ -201,6 +201,12 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.yaml</groupId>
+      <artifactId>snakeyaml</artifactId>
+      <scope>test</scope>
+    </dependency>
+
   </dependencies>
 
   <build>

--- a/app/connector/rest-swagger/src/main/java/io/syndesis/connector/rest/swagger/SpecificationResourceCustomizer.java
+++ b/app/connector/rest-swagger/src/main/java/io/syndesis/connector/rest/swagger/SpecificationResourceCustomizer.java
@@ -20,19 +20,22 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 
-import com.fasterxml.jackson.databind.JsonNode;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import com.fasterxml.jackson.databind.node.ArrayNode;
 import io.syndesis.connector.support.util.ConnectorOptions;
 import io.syndesis.integration.component.proxy.ComponentProxyComponent;
 import io.syndesis.integration.component.proxy.ComponentProxyCustomizer;
+
 import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
 
 public final class SpecificationResourceCustomizer implements ComponentProxyCustomizer {
 
@@ -62,55 +65,71 @@ public final class SpecificationResourceCustomizer implements ComponentProxyCust
     }
 
     /**
-     * Scan the specification and leave only the security method picked by user, if any present.
-     * The change is required to avoid exposure of security configuration (such as apikey) through other channels not requested by user,
-     * for example, query parameters when the user only wants to provide api key through http headers.
-     *
-     * @param specification                    the swagger/openapi original specification
-     * @param securityDefinitionSelectedByUser the securityDefinition picked by the user
-     * @return the updated swagger/openapi specification or the original specification
-     */
-    private static String updateSecuritySpecification(final String specification, final String securityDefinitionSelectedByUser) throws IOException {
-        final Optional<String> securityDefinitionName = getNameFromDefinition(securityDefinitionSelectedByUser);
-        if (securityDefinitionName.isPresent()) {
-            JsonNode rootNode = OBJECT_MAPPER.readTree(specification);
-            List<JsonNode> securities = rootNode.findValues("security");
-            if (!securities.isEmpty()) {
-                LOG.info("Updating specification to accept only {} user selected security method", securityDefinitionName.get());
-                for (JsonNode endpointSecurity : securities) {
-                    if (endpointSecurity.isArray()) {
-                        ArrayNode securityArray = (ArrayNode) endpointSecurity;
-                        for (int i = 0; i < securityArray.size(); i++) {
-                            String securityName = securityArray.get(i).fieldNames().next();
-                            if (!securityDefinitionName.get().equals(securityName)) {
-                                securityArray.remove(i);
-                            }
-                        }
-                    } else {
-                        // Security section must be an array
-                        throw new IllegalArgumentException("Swagger/OpenAPI specification requires endpoint security to be an array of elements!");
-                    }
-                }
-                return OBJECT_MAPPER.writeValueAsString(rootNode);
-            }
-        }
-
-        // no changes
-        LOG.debug("Specification was provided with no security method");
-        return specification;
-    }
-
-    /**
      * Get the name from the security definition or an empty value
      *
-     * @param securityDefinitionSelectedByUser with format type:name (ie, apiKey: api-key-security)
+     * @param securityDefinitionSelectedByUser with format type:name (ie,
+     *            apiKey: api-key-security)
      * @return the name of the definition or an empty value
      */
-    private static Optional<String> getNameFromDefinition(String securityDefinitionSelectedByUser) {
+    private static Optional<String> getNameFromDefinition(final String securityDefinitionSelectedByUser) {
         if (securityDefinitionSelectedByUser != null && securityDefinitionSelectedByUser.indexOf(':') > 0) {
             return Optional.of(securityDefinitionSelectedByUser.substring(securityDefinitionSelectedByUser.indexOf(':') + 1).trim());
         }
 
         return Optional.empty();
+    }
+
+    /**
+     * Scan the specification and leave only the security method picked by user,
+     * if any present. The change is required to avoid exposure of security
+     * configuration (such as apikey) through other channels not requested by
+     * user, for example, query parameters when the user only wants to provide
+     * api key through http headers.
+     *
+     * @param specification the swagger/openapi original specification
+     * @param securityDefinitionSelectedByUser the securityDefinition picked by
+     *            the user
+     * @return the updated swagger/openapi specification or the original
+     *         specification
+     */
+    private static String updateSecuritySpecification(final String specification, final String securityDefinitionSelectedByUser) throws IOException {
+        final Optional<String> securityDefinitionName = getNameFromDefinition(securityDefinitionSelectedByUser);
+        if (!securityDefinitionName.isPresent()) {
+            // no changes
+            LOG.debug("Specification was provided with no security method");
+            return specification;
+        }
+
+        final JsonNode rootNode = OBJECT_MAPPER.readTree(specification);
+        final List<JsonNode> securities = rootNode.findValues("security");
+        if (securities.isEmpty()) {
+            // no changes
+            LOG.debug("Specification was provided with no security method");
+            return specification;
+        }
+
+        LOG.info("Updating specification to accept only {} user selected security method", securityDefinitionName.get());
+        for (final JsonNode endpointSecurity : securities) {
+            if (endpointSecurity.isArray()) {
+                final ArrayNode securityArray = (ArrayNode) endpointSecurity;
+                for (int i = 0; i < securityArray.size(); i++) {
+                    final JsonNode securityDefinition = securityArray.get(i);
+                    final Iterator<String> fields = securityDefinition.fieldNames();
+                    if (!fields.hasNext()) {
+                        continue;
+                    }
+
+                    final String securityName = fields.next();
+                    if (!securityDefinitionName.get().equals(securityName)) {
+                        securityArray.remove(i);
+                    }
+                }
+            } else {
+                // Security section must be an array
+                throw new IllegalArgumentException("Swagger/OpenAPI specification requires endpoint security to be an array of elements!");
+            }
+        }
+
+        return OBJECT_MAPPER.writeValueAsString(rootNode);
     }
 }

--- a/app/connector/rest-swagger/src/test/java/io/syndesis/connector/rest/swagger/SpecificationResourceCustomizerTest.java
+++ b/app/connector/rest-swagger/src/test/java/io/syndesis/connector/rest/swagger/SpecificationResourceCustomizerTest.java
@@ -20,47 +20,21 @@ import java.io.IOException;
 import java.util.HashMap;
 import java.util.Map;
 
+import io.syndesis.common.util.json.JsonUtils;
 import io.syndesis.connector.support.util.ConnectorOptions;
 import io.syndesis.integration.component.proxy.ComponentProxyComponent;
+
 import org.apache.camel.util.StringHelper;
 import org.junit.jupiter.api.Test;
+import org.yaml.snakeyaml.Yaml;
+
+import com.fasterxml.jackson.databind.JsonNode;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class SpecificationResourceCustomizerTest {
 
     private static final ComponentProxyComponent NOT_USED = null;
-
-    @Test
-    public void shouldStoreSpecificationInTemporaryDirectory() {
-        final SpecificationResourceCustomizer customizer = new SpecificationResourceCustomizer();
-
-        final Map<String, Object> options = new HashMap<>();
-        options.put("specification", "the specification is here");
-
-        customizer.customize(NOT_USED, options);
-
-        assertThat(options).containsKey("specificationUri");
-        assertThat(options).doesNotContainKey("specification");
-        assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent("the specification is here");
-    }
-
-    @Test
-    public void shouldRemoveUnwantedSecurity() throws IOException {
-        final SpecificationResourceCustomizer customizer = new SpecificationResourceCustomizer();
-        final String spec = RestSwaggerConnectorIntegrationTest.readSpecification("apikey.json");
-        final String specUpdated = RestSwaggerConnectorIntegrationTest.readSpecification("apikey-security-updated.json");
-
-        final Map<String, Object> options = new HashMap<>();
-        options.put("specification", spec);
-        options.put("authenticationType", "apiKey: api-key-header");
-
-        customizer.customize(NOT_USED, options);
-
-        assertThat(options).containsKey("specificationUri");
-        assertThat(options).doesNotContainKey("specification");
-        assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent(specUpdated);
-    }
 
     @Test
     public void shouldNotUpdateSpecificationOnMissingSecurityDefinitionName() throws IOException {
@@ -90,6 +64,52 @@ public class SpecificationResourceCustomizerTest {
 
         assertThat(options).containsKey("specificationUri");
         assertThat(options).doesNotContainKey("specification");
+        assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent(spec);
+    }
+
+    @Test
+    public void shouldRemoveUnwantedSecurity() throws IOException {
+        final SpecificationResourceCustomizer customizer = new SpecificationResourceCustomizer();
+        final String spec = RestSwaggerConnectorIntegrationTest.readSpecification("apikey.json");
+        final String specUpdated = RestSwaggerConnectorIntegrationTest.readSpecification("apikey-security-updated.json");
+
+        final Map<String, Object> options = new HashMap<>();
+        options.put("specification", spec);
+        options.put("authenticationType", "apiKey: api-key-header");
+
+        customizer.customize(NOT_USED, options);
+
+        assertThat(options).containsKey("specificationUri");
+        assertThat(options).doesNotContainKey("specification");
+        assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent(specUpdated);
+    }
+
+    @Test
+    public void shouldStoreSpecificationInTemporaryDirectory() {
+        final SpecificationResourceCustomizer customizer = new SpecificationResourceCustomizer();
+
+        final Map<String, Object> options = new HashMap<>();
+        options.put("specification", "the specification is here");
+
+        customizer.customize(NOT_USED, options);
+
+        assertThat(options).containsKey("specificationUri");
+        assertThat(options).doesNotContainKey("specification");
+        assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent("the specification is here");
+    }
+
+    @Test // issue #9411
+    void shouldNotFailToUpdateSpecificationWithApiKeyAuthEmptyArray() {
+        final SpecificationResourceCustomizer customizer = new SpecificationResourceCustomizer();
+        final String specYaml = RestSwaggerConnectorIntegrationTest.readSpecification("twitter-api.yaml");
+        final JsonNode jsonNode = JsonUtils.convertValue(new Yaml().load(specYaml), JsonNode.class);
+        final String spec = JsonUtils.toString(jsonNode);
+
+        final Map<String, Object> options = new HashMap<>();
+        options.put("specification", spec);
+        options.put("authenticationType", "apiKey: ApiKeyAuth");
+
+        customizer.customize(NOT_USED, options);
         assertThat(new File(StringHelper.after(ConnectorOptions.extractOption(options, "specificationUri"), "file:"))).hasContent(spec);
     }
 }

--- a/app/connector/rest-swagger/src/test/resources/twitter-api.yaml
+++ b/app/connector/rest-swagger/src/test/resources/twitter-api.yaml
@@ -1,0 +1,46 @@
+openapi: 3.0.2
+info:
+  title: Tweeter API - GET
+  description: some description
+  version: 0.1.9
+servers:
+  - url: https://api.twitter.com
+    description: Optional server description, e.g. Main (production) server
+paths:
+  /labs/2/users/{id}:
+    get:
+      summary: Get types.
+      parameters:
+        - name: id
+          in: path
+          description: User ID
+          required: true
+          schema:
+            type: integer
+            format: int64
+      responses:
+        '200':
+          description: A list of types.
+          content:
+            application/json:
+              schema:
+                  $ref: '#/components/schemas/User'
+        '500':
+          description: Error
+components:
+  schemas:
+    User:
+      properties:
+        id:
+          type: string
+        name:
+          type: string
+        username:
+          type: string
+  securitySchemes:
+    ApiKeyAuth:
+      type: apiKey
+      in: header
+      name: Authorization
+security:
+- ApiKeyAuth: []

